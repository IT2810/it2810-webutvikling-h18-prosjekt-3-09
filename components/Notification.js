// import Snackbar from 'react-native-snackbar';

export const sendNotification = (type, message) => {
  let backgroundColor = "darkgrey";
  switch (type) {
    case "success":
      backgroundColor = "green";
      break;
    case "warning":
      backgroundColor = "gold";
      break;
    case "error":
      backgroundColor = "darkred";
      break;

    default:
      break;
  }

  /*
   * TODO: Fix notification
   * Snackbar.show({
   *   title: message || "Message",
   *   duration: Snackbar.LENGTH_LONG,
   *   backgroundColor
   * });
   */
  return { backgroundColor, message };
};
