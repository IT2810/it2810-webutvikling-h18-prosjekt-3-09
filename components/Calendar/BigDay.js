import React, { Component } from "react";
import moment from "moment";

import { ScrollView, View } from "react-native";
import DateTimePicker from "react-native-modal-datetime-picker";
import { Col, Grid } from "react-native-easy-grid";
import { sendNotification } from "../Notification";
import { ID, validateEvent } from "../../utils";
import colors from "../../constants/Colors";
import {
  Button,
  Card,
  List,
  ListItem,
  FormLabel,
  FormInput,
  Header,
  Text
} from "react-native-elements";

import DateInput from "./DateInput";
const initialEvent = {
  id: "",
  start: moment().toDate(),
  end: moment()
    .add(1, "hour")
    .toDate(),
  title: ""
};

export default class extends Component {
  state = {
    dateType: "start",
    isDateTimePickerVisible: false,
    ...initialEvent
  };

  showDateTimePicker = dateType =>
    this.setState({ isDateTimePickerVisible: true, dateType });

  hideDateTimePicker = () => this.setState({ isDateTimePickerVisible: false });

  handleDatePicked = date => {
    this.setState({ [this.state.dateType]: date });
    this.hideDateTimePicker();
  };

  handleTitleChange = title => this.setState({ title });

  reset = () => this.setState({ ...initialEvent });

  handleSave = () => {
    const { start, end, title, id } = this.state;

    const isNewEvent = id === "";
    let newEvent = { title, start, end };
    if (isNewEvent) {
      newEvent = {
        ...newEvent,
        id: ID()
      };
    } else {
      newEvent = { ...newEvent, id };
    }

    const invalid = validateEvent(newEvent);
    if (!invalid) {
      if (isNewEvent) {
        this.props.createEvent(newEvent);
      } else {
        this.props.changeEvent(newEvent);
      }
      this.reset();
      sendNotification("success", "Event is created");
    } else {
      sendNotification("error", invalid);
    }
  };

  handleDelete = id => {
    this.props.deleteEvent(id);
    this.reset();
  };

  setEditedEvent = eventId => {
    const { start, end, title } = this.props.events.find(
      ({ id }) => id === eventId
    );
    this.setState({ id: eventId, start, end, title });
  };

  render() {
    const { start, end, dateType, title, id } = this.state;
    const { value, events } = this.props;

    const isNew = id === "";

    return (
      <ScrollView>
        <Header
          backgroundColor={colors.color1}
          centerComponent={{
            text: moment(value).format("MMM DD."),
            style: { color: colors.light }
          }}
          leftComponent={{
            icon: "arrow-back",
            color: "#fff",
            onPress: () => this.props.changeDay(-1)
          }}
          rightComponent={{
            icon: "arrow-forward",
            color: "#fff",
            onPress: () => this.props.changeDay(1)
          }}
        />
        <Grid style={{ marginTop: 16 }}>
          <Col>
            <Button
              backgroundColor={colors.color1}
              onPress={this.handleSave}
              title={isNew ? "Create" : "Save"}
            />
          </Col>
          <Col>
            <Button
              backgroundColor={colors.color4}
              onPress={() =>
                isNew ? this.props.handleClose() : this.handleDelete(id)
              }
              title={isNew ? "Close" : "Delete"}
            />
          </Col>
        </Grid>
        <Card title={`${isNew ? "Create" : "Edit"} event`}>
          <View style={{ flexDirection: "row", marginBottom: 8 }}>
            <FormLabel>Title</FormLabel>
            <FormInput
              inputStyle={{ borderColor: "black", borderBottomWidth: 1 }}
              placeholder="Enter title"
              onChangeText={this.handleTitleChange}
              value={title}
            />
          </View>
          <DateInput
            label="Start date"
            onPress={() => this.showDateTimePicker("start")}
            value={start}
          />
          <DateInput
            label="End date"
            onPress={() => this.showDateTimePicker("end")}
            value={end}
          />
        </Card>
        <Card title="Events on this day:">
          {events.length ? (
            <List>
              {events.map(({ id, title, start, end }) => (
                <ListItem
                  title={title}
                  key={title}
                  subtitle={`${moment(start).format("YYYY-MM-DD")} â€¢ ${moment(
                    end
                  ).format("YYYY-MM-DD")}`}
                  onPress={() => this.setEditedEvent(id)}
                />
              ))}
            </List>
          ) : (
            <Text>No events for today</Text>
          )}
        </Card>
        <DateTimePicker
          date={moment(
            (dateType === "start" ? start : end) || undefined
          ).toDate()}
          isVisible={this.state.isDateTimePickerVisible}
          onConfirm={this.handleDatePicked}
          onCancel={this.hideDateTimePicker}
        />
      </ScrollView>
    );
  }
}
