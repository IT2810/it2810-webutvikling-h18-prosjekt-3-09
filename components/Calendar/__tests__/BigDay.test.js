import React from "react";
import BigDay from "../BigDay";
import renderer from "react-test-renderer";
import Moment from "moment";
import { extendMoment } from "moment-range";
import * as notification from "../../Notification";

const moment = extendMoment(Moment);

describe("BigDay snapshot.", () => {
  const initialEvent = {
    id: "",
    start: moment().toDate(),
    end: moment()
      .add(1, "hour")
      .toDate(),
    title: ""
  };

  const invalidID = null;

  const testEvent = {
    id: "0000",
    start: moment()
      .add(1, "week")
      .toDate(),
    end: moment()
      .add(2, "week")
      .toDate(),
    title: "Test event"
  };

  let tree;
  beforeAll(() => {
    tree = renderer.create(
      <BigDay
        events={[testEvent]}
        deleteEvent={jest.fn()}
        createEvent={jest.fn()}
        changeEvent={jest.fn()}
      />
    );
  });
  test("renders correctly", () => {
    expect(tree).toMatchSnapshot();
  });

  test("events are received through props", () => {
    expect(tree.toTree().props.events.length).toBe(1);
  });

  describe("DateTimePicker visibility", () => {
    test("DateTimePicker gets visible", () => {
      tree.toTree().instance.showDateTimePicker("start");
      expect(tree.toTree().instance.state.isDateTimePickerVisible).toBe(true);
    });

    test("DateTimePicker gets hidden", () => {
      tree.toTree().instance.hideDateTimePicker();
      expect(tree.toTree().instance.state.isDateTimePickerVisible).toBe(false);
    });

    test("end DatePicker gets visible when picking 'end'", () => {
      tree.toTree().instance.showDateTimePicker("end");
      expect(tree.toTree().instance.state.dateType).toBe("end");
      expect(tree.toTree().instance.state.isDateTimePickerVisible).toBe(true);
    });
  });

  describe("Event handling", () => {
    beforeAll(() => {
      tree.toTree().instance.setEditedEvent(testEvent.id);
    });
    describe("Set event that is edited", () => {
      test("id is set", () => {
        expect(tree.toTree().instance.state.id).toBe(testEvent.id);
      });
      test("title is set", () => {
        expect(tree.toTree().instance.state.id).toBe(testEvent.id);
      });
      test("start date is set", () => {
        moment(tree.toTree().instance.state.start).isSame(
          moment(testEvent.start),
          "day"
        );
      });
      test("end date is set", () => {
        moment(tree.toTree().instance.state.end).isSame(
          moment(testEvent.end),
          "day"
        );
      });
    });
    describe("Change event", () => {
      test("changing title", () => {
        tree.toTree().instance.handleTitleChange(testEvent.title);
        expect(tree.toTree().instance.state.title).toBe(testEvent.title);
      });

      test("changing start date", () => {
        tree.toTree().instance.showDateTimePicker("start");
        tree.toTree().instance.handleDatePicked(testEvent.start);
        tree.toTree().instance.hideDateTimePicker();
        expect(tree.toTree().instance.state.start).toBe(testEvent.start);
      });

      test("changing end date", () => {
        tree.toTree().instance.showDateTimePicker("end");
        tree.toTree().instance.handleDatePicked(testEvent.end);
        tree.toTree().instance.hideDateTimePicker();
        expect(tree.toTree().instance.state.end).toBe(testEvent.end);
      });
    });

    describe("Reset event", () => {
      test("id is reset", () => {
        tree.toTree().instance.state.id = testEvent.id;
        tree.toTree().instance.reset();
        expect(tree.toTree().instance.state.id).toBe(initialEvent.id);
      });
      test("title is reset", () => {
        tree.toTree().instance.state.title = testEvent.title;
        tree.toTree().instance.reset();
        expect(tree.toTree().instance.state.title).toBe(initialEvent.title);
      });
      test("start date is reset", () => {
        tree.toTree().instance.state.start = testEvent.start;
        tree.toTree().instance.reset();
        expect(
          moment(tree.toTree().instance.state.start).isSame(
            moment(initialEvent.start),
            "day"
          )
        ).toBe(true);
      });
      test("end date is reset", () => {
        tree.toTree().instance.state.end = testEvent.end;
        tree.toTree().instance.reset();
        expect(
          moment(tree.toTree().instance.state.end).isSame(
            moment(initialEvent.end),
            "day"
          )
        ).toBe(true);
      });
    });

    describe("Save event", () => {
      beforeEach(() => {
        tree.toTree().instance.state.title = testEvent.title;
        tree.toTree().instance.state.start = testEvent.start;
        tree.toTree().instance.state.end = testEvent.end;
      });
      test("new event created", () => {
        tree.toTree().instance.state.id = "";
        tree.toTree().instance.handleSave();
        expect(tree.toTree().instance.props.createEvent).toBeCalled();
      });
      test("event change was saved", () => {
        tree.toTree().instance.state.id = testEvent.id;
        tree.toTree().instance.handleSave();
        expect(tree.toTree().instance.props.changeEvent).toBeCalled();
      });
      test("invalid event sends notification", () => {
        notification.sendNotification = jest.fn();
        tree.toTree().instance.state.id = invalidID;
        tree.toTree().instance.handleSave();
        expect(notification.sendNotification).toBeCalled();
      });
    });

    describe("Delete event", () => {
      test("event gets deleted", () => {
        tree.toTree().instance.state.id = testEvent.id;
        tree.toTree().instance.handleDelete(testEvent.id);
        expect(tree.toTree().instance.state.id).toBe(initialEvent.id);
      });
    });
  });
});
