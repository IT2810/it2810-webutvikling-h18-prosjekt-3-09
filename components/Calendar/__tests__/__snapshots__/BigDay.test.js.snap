// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigDay snapshot. mounting  1`] = `null`;

exports[`BigDay snapshot. renders correctly 1`] = `null`;

exports[`BigDay snapshot. updating  1`] = `null`;
